verification/spec/connection-handshake/ConnectionHandshakeModule.tla

```diff
@@ -0,0 +1,353 @@
+    outBuf
+
+
+moduleVars == <<inBuf, outBuf, store>>
+
+
+(***************************************************************************
+    Helper operators.
+ ***************************************************************************)
+
+
+(* Returns true if 'para' matches the parameters in the local connection,
+    and returns false otherwise.
+ *)
+CheckLocalParameters(para) ==
+    LET local == store.connection.parameters.localEnd
+        remote == store.connection.parameters.remoteEnd 
+    IN /\ local.connectionID = para.localEnd.connectionID   
+       /\ remote.connectionID = para.remoteEnd.connectionID
+       /\ local.clientID = para.localEnd.clientID       
+       /\ remote.clientID = para.remoteEnd.clientID
+
+
+(* Validates a ConnectionParameter.
+    Expects as input a ConnectionParameter 'para' and returns true or false.
+    
+    This is a basic validation step, making sure that 'para' is valid with
+    respect to module-level constants ConnectionIDs and ClientIDs.
+    If there is a connection in the store, this also validates 'para'
+    against the parameters of an existing connection by relying on the
+    state predicate CheckLocalParameters.
+*)
+ValidConnectionParameters(para) ==
+    /\ para.localEnd.connectionID \in ConnectionIDs
+    /\ para.localEnd.clientID \in ClientIDs
+    /\ \/ store.connection = NullConnection
+       \/ /\ store.connection /= NullConnection
+          /\ CheckLocalParameters(para)
+
+
+(* Operator for reversing the connection ends.
+
+    Given a ConnectionParameters record 'para', returns a new set
+    of parameters where the local and remote ends are
+    flipped (i.e., reversed).
+ *)
+FlipConnectionParameters(para) ==
+    [localEnd   |-> para.remoteEnd,
+     remoteEnd  |-> para.localEnd]
+
+
+(* Operator for construcing a connection proof.
+
+    The connection proof is used to demonstrate to another chain that the
+    local store on this chain comprises a connection in a certain state. 
+ *)
+GetConnProof(connState) ==
+    [height |-> store.latestHeight,
+     connectionState |-> connState]
+
+
+(* Operator for construcing a client proof.
+ *)
+GetClientProof ==
+    [height |-> store.client.latestHeight]
+
+
+(* Verification of a connection proof. 
+
+    This is a state predicate returning true if the following holds: the local
+    client on this chain stores the height reported in the proof. Note that this
+    condition should eventually become true, assuming a correct environment, which
+    should periodically update the client on each chain; see actions 'GoodNextEnv'
+    and 'UpdateClient'.
+
+ *)
+VerifyConnProof(cp) ==
+    /\ cp.height \in store.client.consensusStates
+
+
+(* Verification of a client proof.
+
+    This is a state predicate returning true if the following holds: the height
+    reported in the client proof must not exceed the current (latestHeight) of
+    this chain.
+ *)
+VerifyClientProof(cp) ==
+    cp.height <= store.latestHeight
```

Comment:
Here we miss the check that the corresponding height is installed at the destination.

Reply:
This is going to be weird in the current model as it would mean looking up the remote chain.

Reply:
Per @ancazamfir suggestion, the proof now includes the set of all heights, and added a verification against that set.